chore: remove organization_id suffix from org_member roles in database

Emyrk opened the pull request:
**Any existing organization role assignments are wiped. No one should be using these**

Organization member's table is already scoped to an organization.
Rolename should avoid having the org_id appended.

When adding custom roles, this makes more sense to keep role names consistent across different tables. The only awkward thing is that in `coderdtest` we pass site and org roles as strings. So these strings still need `rolename:<org_id>` for the function to know which roles are org scoped, and which are site for assigning. This is only in tests though, which can use the rbac rolename helper functions.


I did change the inputs from `organization-admin:432475a4-6ca2-4a33-b8f7-2b26e8aa729d` -> `organization-admin`. I added a new error to tell the caller what the change they need to make it. We only used this in unit tests. No UI, no cli, and no one should be using this endpoint, so it is not a breaking change.

---

The custom roles effort is adding `OrganizationID` as a field on orgs, rather than a naming convention.

Emyrk committed:
chore: refactor, remove organiation_id from org_member roles

Organization member's table is already scoped to an organization.
Rolename should avoid having the org_id appended

Emyrk committed:
add safety check

Emyrk committed:
SlimRole has organizaton scope now in response, remove from the name

Emyrk committed:
make gen

Emyrk committed:
Linting

Emyrk committed:
fix test

Emyrk commented on coderd/database/dbauthz/dbauthz.go:
```diff
@@ -2847,8 +2847,22 @@ func (q *querier) UpdateMemberRoles(ctx context.Context, arg database.UpdateMemb
+	// Convert the argument roles for validation.
+	scopedGranted := make([]string, 0, len(arg.GrantedRoles))
+	for _, grantedRole := range arg.GrantedRoles {
+		// This check is a developer safety check. Old code might try to invoke this code path with
+		// organization id suffixes. Catch this and return a nice error so it can be fixed.
+		_, foundOrg, _ := rbac.RoleSplit(grantedRole)
+		if foundOrg != "" {
+			return database.OrganizationMember{}, xerrors.Errorf("attempt to assign a role %q, remove the ':<organization_id> suffix", grantedRole)
+		}
```
This is not required, but added as a safety check to catch dev errors from using the previous `rolename:<org_id>` syntax

johnstcn commented on coderd/database/migrations/000215_scoped_org_db_roles.up.sql:
```diff
@@ -0,0 +1,7 @@
+-- The default was 'organization-member', but we imply that in the
+-- 'GetAuthorizationUserRoles' query.
+ALTER TABLE ONLY organization_members ALTER COLUMN roles SET DEFAULT '{}';
+
+-- No one should be using organization roles yet. If they are, the names in the
+-- database are now incorrect. Just remove them all.
+UPDATE organization_members SET roles = '{}';
```
another good thing to call out

johnstcn commented on coderd/rbac/roles.go:
```diff
@@ -70,6 +70,14 @@ func RoleMember() string {
 	return RoleName(member, "")
 }
 
+func StaticRoleOrgAdmin() string {
+	return orgAdmin
+}
+
+func StaticRoleOrgMember() string {
+	return orgMember
+}
+
```
What's the difference between "Static" org admin and `RoleOrgAdmin()`?

Emyrk replied:
This is super annoying, but essentially I should make `RoleOrgAdmin(<org_id>)` not take an organization ID. But there is >80 usages of `RoleOrgAdmin()` and `RoleOrgMember()`. All our test APIs just take `[]string{}` for roles, but now they should take, `[]string, map[<org_id>][]string`.

That kind of refactor would be a much larger diff, and this would only serve to affect unit tests.

Emyrk replied:
I have another idea. I will rename the old ones to something else and put up a deprecated comment.

Emyrk replied:
Solution

https://github.com/coder/coder/blob/cf933079a969cda57af4d51ae0e9a4c1aff50d80/coderd/rbac/roles.go#L73-L94

Emyrk committed:
chore: deprecate scoped org role names from the rbac package

Fixing all the test apis to remove this is a lot of work for little
return atm.

johnstcn commented on docs/api/enterprise.md:
```diff
@@ -1600,7 +1600,8 @@ curl -X PATCH http://coder-server:8080/api/v2/scim/v2/Users/{id} \
       "display_name": "string",
-      "name": "string"
+      "name": "string",
+      "organization_id": "string"
```
We should probably also drop a comment here regarding the experimental status of this thing until it's ready for prime time

Emyrk replied:
I wish the docs would read the `omitempty` field like our type generator does. No organization roles are currently returned by any api in use by our frontend, so the field is always excluded in the json responses.

johnstcn reviewed (approved):
LGTM, but we should have a follow-up to remove the usage of these deprecated `ScopedRole.*` functions. Also it might be no harm to have some comments on the `organization_id` fields in our API docs regarding the current experimental status.
